### ISSUE TITLE
Hierarchical Classification in Yolo v8 

### ISSUE THREAD

**Issue by bcwein** [ultralytics/ultralytics#4353]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello

I have created my own new custom dataset of annotated images for instance segmentation of marine life. Given the hierarchical nature of taxonomy and existing domain knowledge in the field, I would like to experiment with hierarchical classification where the classes are related in a tree-like structure. For example, predicting a higher level class if the confidence is low. 

1. Is this idea feasible? 
2. How could I modify the model structure to incorporate hierarchical classification? (e.g. modify the classification head)

**Comment by GrunCrow** [comment]
Hi!

I'm actually trying to develop a very similar model. I'm doing camera trap mammals classification and I would like to do a hierarchical classification just as you explained. I also have my custom dataset. My idea is finding for each bounding box all the classes confidence and then from here try to build the hierarchical classification thing.

For the moment I haven't found a solution, but I found this other issues than gave me some ideas of where to look or what to look for, maybe it can help you too. If I do any other improvement or any progress I will let you know :)

[All class confidence](https://github.com/ultralytics/yolov5/issues/4122) - This one is from Yolov5
[Box Confidences and Class Probabilities](https://github.com/ultralytics/ultralytics/issues/2423)

**Comment by pderrenger** [comment]
@GrunCrow hi there!

Your project sounds fascinating! 👏 Indeed, handling hierarchical classification in YOLOv8 can be quite intriguing. To tackle this, you can modify the classification head of the model to incorporate the hierarchical structure. One way to approach this is to design a branching architecture that predicts classes at different levels of the hierarchy based on the confidence scores.

However, it's important to note that this modification may require a deeper understanding of the model's architecture and possibly some experimentation to find the most effective approach for your specific use case.

The issues you found in YOLOv5 and within the Ultralytics repo contain valuable insights that can provide useful guidance for your project. As you make progress, feel free to share any developments or improvements, and I'd be happy to help further!

Happy coding! 🚀

**Comment by GrunCrow** [comment]
> @GrunCrow hi there!
> 
> Your project sounds fascinating! 👏 Indeed, handling hierarchical classification in YOLOv8 can be quite intriguing. To tackle this, you can modify the classification head of the model to incorporate the hierarchical structure. One way to approach this is to design a branching architecture that predicts classes at different levels of the hierarchy based on the confidence scores.
> 
> However, it's important to note that this modification may require a deeper understanding of the model's architecture and possibly some experimentation to find the most effective approach for your specific use case.
> 
> The issues you found in YOLOv5 and within the Ultralytics repo contain valuable insights that can provide useful guidance for your project. As you make progress, feel free to share any developments or improvements, and I'd be happy to help further!
> 
> Happy coding! 🚀

Hi! I've already addressed the issue (a few weeks after my initial message). During a workshop, we developed a [script to extract YOLOv8 NMS and activation points](https://github.com/CV4EcologySchool/yolo-hacks). Following that, I created a [post-processing script](https://github.com/GrunCrow/cv4ecology/tree/main/AI_Census/Hierarchical%20Model) where various parameters, including the temperature of the softmax, threshold, and hierarchy threshold, could be adjusted, leading to a recalculation of the classification. If you're interested, I can provide a more detailed explanation; this is just a brief update. Here you have some examples of classification on test images:

Classification of species (the species score was sufficient to classify it correctly)
![image_2023-11-16_100102394](https://github.com/ultralytics/ultralytics/assets/57197254/d48ff1c7-6b75-4064-9457-8099241e13b7)

Family classification, where the species score wasn't adequate, it checked the family score, successfully avoiding an error in identifying a fallow deer as a red deer. However, further examination by experts confirmed it was indeed a fallow deer. This suggests potential issues with mislabeled images of fallow deer and red deer, which look very similar. Adjustments to the training, validation, and test datasets may be necessary.
![image_2023-11-16_100142459](https://github.com/ultralytics/ultralytics/assets/57197254/fe068191-d00a-4510-9592-8cf0de4bf689)

Another example involving cervids:
![image_2023-11-16_100420096](https://github.com/ultralytics/ultralytics/assets/57197254/7c5fb3c3-5c74-4e9f-8a2b-516206f1cc04)

If the score is insufficient for species or family classification, the system categorizes it as an animal, effectively functioning as a detector:
![image_2023-11-16_100339281](https://github.com/ultralytics/ultralytics/assets/57197254/0418abbd-fdda-457c-bf87-84baead54cf0)

Here is an image to illustrate how the hierarchy functions. Initially, it checks if the species threshold is met; if not, it proceeds to calculate the family score and compares it again. This process repeats until the score surpasses the hierarchy threshold.
![image_2023-11-16_101022291](https://github.com/ultralytics/ultralytics/assets/57197254/03402ea8-673c-4f19-8d67-cc36b605bfbd)

I implemented this mechanism using dictionaries, but I believe it could operate more efficiently and be easier to implement with a tree data structure. Unfortunately, I'm no longer involved with this project and research team, so I'm not actively working on it at the moment. However, I'm enthusiastic about completing it in the near future.

**Comment by glenn-jocher** [comment]
Hi @GrunCrow,

That's incredible progress! 🌟 It sounds like you've made significant headway with your hierarchical classification approach, and the results you've shared are quite promising.

The method you've developed—leveraging a post-processing script to adjust classification based on hierarchical confidence levels—is quite a clever solution. It's fascinating to see the adaptation of the model to not just classify precisely but also to recognize uncertainty and step up the hierarchy to ensure more accurate results. Your approach to avoid incorrect classification by considering family scores when species confidence is low is a smart way to introduce domain knowledge into the model's decision-making process.

Using dictionaries for this purpose is a straightforward initial method, but I agree that a tree data structure might offer better performance and clarity for expansions or future modifications of the hierarchy.

While you might not be currently working on this project, what you've accomplished so far is undoubtedly valuable to the community and could inspire similar solutions in other niche applications of object classification. It’s great to see such pioneering work within the YOLOv8 and broader AI ecosystem.

Keep up the innovative spirit and when you do get around to completing it, I’d be thrilled to see the finished product, as would many others, I’m sure. Keep pushing the boundaries! 🚀💡

Should you or anyone in the community have further questions or need assistance related to YOLOv8 and hierarchical classification, feel free to reach out. Keep fostering that spirit of collaboration and open development!

Best wishes on your next ventures!

**Comment by bcwein** [comment]
@GrunCrow Wow! Incredible progress and very impressive!

I would love to hear more about this and try to incorporate this into my own work. I am currently busy with other projects for a little while but will after a couple of weeks resume my work on hierarchical prediction using taxonomy.

**Comment by GrunCrow** [comment]
@bcwein Great, if you need any help I would love to help you! Just text me here, on twitter or wherever you prefer and we could talk more about it

(Tw: @GrunCrow)

**Comment by bcwein** [comment]
@Gr2unCrow Thanks for your script! It has helped get me started on hierarchical classification for my use case. I would also love to meet to discuss further.

I have a question regarding what exactly "activations" is in the [hierarcical classification notebook](https://github.com/GrunCrow/cv4ecology/blame/d1f9da99df6d4d4d4322637f6865bbe2cf96bb18/AI_Census/Hierarchical%20Model/c_Hierarchical_Classification.ipynb#L170)

The particular list in that dictionary does not sum up to 1. So it is not probabilities in the formal sense, which is what I want to extract.

In the post-processing script, the NMS function from [ultralytics](https://docs.ultralytics.com/reference/utils/ops/#ultralytics.utils.ops.non_max_suppression) is used [here](https://github.com/CV4EcologySchool/yolo-hacks/blob/df3b58c87a36f6e3682f6f75cf3215f2b11c28f1/predict_with_logits.py#L272C1-L282C10) 

I am not able to fully understand the input tensor format used there. The input format is described as:

"A tensor of shape (batch_size, num_classes + 4 + num_masks, num_boxes) containing the predicted boxes, classes, and masks. The tensor should be in the format output by a model, such as YOLO." in the [documentation](https://docs.ultralytics.com/reference/utils/ops/#ultralytics.utils.ops.non_max_suppression)

Can @pderrenger or someone at Ultralytics help verify if this approach does indeed is on the right track on how to extract all class probabilities?

When running the [function in the post processing script](https://github.com/CV4EcologySchool/yolo-hacks/blob/df3b58c87a36f6e3682f6f75cf3215f2b11c28f1/predict_with_logits.py#L213) on my own trained Yolo model i get:

```python
results[0].boxes[0]
```

`ultralytics.engine.results.Boxes object with attributes:

cls: tensor([0.], device='cuda:0')
conf: tensor([0.8566], device='cuda:0')
data: tensor([[6.9781e+02, 4.7114e+02, 8.6375e+02, 5.6334e+02, 8.5660e-01, 0.0000e+00]], device='cuda:0')
id: None
is_track: False
orig_shape: (1080, 1920)
shape: torch.Size([1, 6])
xywh: tensor([[780.7778, 517.2428, 165.9397,  92.2033]], device='cuda:0')
xywhn: tensor([[0.4067, 0.4789, 0.0864, 0.0854]], device='cuda:0')
xyxy: tensor([[697.8080, 471.1412, 863.7477, 563.3445]], device='cuda:0')
xyxyn: tensor([[0.3634, 0.4362, 0.4499, 0.5216]], device='cuda:0')
`


```python
boxes[0]
```

`
{'bbox': [697.8079833984375,
  471.1411437988281,
  863.7476806640625,
  563.344482421875],
 'bbox_xywh': [780.77783203125,
  517.2428131103516,
  165.939697265625,
  92.20333862304688],
 'best_conf': 0.85660320520401,
 'best_cls': 0.0,
 'image_id': '/mnt/RAID/datasets/label-studio/fjord/images/frame003206.jpg',
 'activations': [-1.548038125038147,
  2.4233124256134033,
  0.7758280634880066,
  0.2959475517272949,
  -0.07892662286758423,
  -0.0406077541410923,
  0.06858810782432556,
  0.9491586685180664,
  -1.1427911520004272,
  0.28358930349349976,
  0.7851924896240234,
  -1.3418079614639282,
  0.11901772022247314],
 'logits': [0.13132593035697937,
...
  -11.774755477905273,
  -12.661630630493164,
  -14.265462875366211,
  -13.874723434448242,
  -13.744763374328613]}
  
 I am unsure how to proceed with the activations extracted for hierarchical classification. 'logits' is also a very long list way longer than number of classes.

**Comment by GrunCrow** [comment]
Hi!

Long time no dive into that hierarchical stuff, but I'll try to recall the best I can. Just so you know, the techy magic in the script was mostly done by [@justinkay](https://github.com/justinkay), so he might have more insights.

Figuring out what those boxes held was confusing for me as well. As I said it was months ago so it's hard to me to remember all what I took from the experiments.

Checked out your output, and it's giving off some weird vibes. In my case I saved it in a [json](https://github.com/GrunCrow/cv4ecology/blob/d1f9da99df6d4d4d4322637f6865bbe2cf96bb18/new_predictions.json) but the activation scores look much better than yours, they do not sum 1 as I just checked, but it is close to 1. Blame it on computers rounding off those decimals. In this image I calculated with python the sum of one of the activations values I got:
![image](https://github.com/ultralytics/ultralytics/assets/57197254/876beec6-6510-4963-aaa1-d9eca7c1be6b)

On the other hand, from what I understood, I consider that first you need to "know" the logits, those are the output of the neural network which are numbers that says almost nothing to us. Then we apply an activation function to try to make them more usable from the human perspective. Which could be "probabilities". What the activation function does is a weighted sum. 
The thing that I learnt in the workshop [cv4ecology ](https://cv4ecology.caltech.edu) (if they open a new edition, I recommend you to apply) is that they are real probabilities so we should stop calling and using them as probabilities, they are more like score points. Now there is no need to worry about them summing up 1 or not.

So maybe some weird thing is happening between your original yolo predictions and the activations and nms functions. Maybe if you could share the code you are using.

How many classes are you training YOLO with? How is your hierarchy?

Also, I would check YOLO and libraries versions. YOLO is constantly changing and maybe an important part has changed since the script was developed and is affecting how it is calculated.

Next weeks are going to be very busy weeks for me, but we can try to schedule a meeting to discuss this further and see if I can assist you in any way.
